File: java/version1_soap.java
// http://localhost:8080/?n=10
// javac -cp . version1_soap.java
// java -cp . version1_soap

import com.sun.net.httpserver.HttpServer;
import java.io.*;
import java.net.*;
import java.net.http.*;

public class version1_soap {
    public static void main(String[] args) throws Exception {
        HttpServer server = HttpServer.create(new InetSocketAddress(8080), 0);
        server.createContext("/", exchange -> {
            try {
                String query = exchange.getRequestURI().getQuery();
                String n = "0";
                if (query != null && query.contains("n=")) {
                    n = query.split("n=")[1];
                }

                String soap = "<?xml version=\"1.0\" encoding=\"utf-8\"?>"
                    + "<soap:Envelope xmlns:soap=\"http://schemas.xmlsoap.org/soap/envelope/\">"
                    + "<soap:Body>"
                    + "<NumberToWords xmlns=\"http://www.dataaccess.com/webservicesserver/\">"
                    + "<ubiNum>" + n + "</ubiNum>"
                    + "</NumberToWords>"
                    + "</soap:Body>"
                    + "</soap:Envelope>";

                HttpClient client = HttpClient.newHttpClient();
                HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create("https://www.dataaccess.com/webservicesserver/NumberConversion.wso"))
                    .header("Content-Type", "text/xml; charset=utf-8")
                    .POST(HttpRequest.BodyPublishers.ofString(soap))
                    .build();

                HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
                String body = response.body();

                String result;
                if (body.contains("NumberToWordsResult>")) {
                    result = body.split("NumberToWordsResult>")[1].split("<")[0];
                } else {
                    result = "Error: " + body.substring(0, 200);
                }

                byte[] resp = result.getBytes("UTF-8");
                exchange.getResponseHeaders().set("Content-Type", "text/plain; charset=UTF-8");
                exchange.sendResponseHeaders(200, resp.length);
                exchange.getResponseBody().write(resp);
                exchange.getResponseBody().close();

            } catch (Exception e) {
                byte[] resp = ("Error: " + e.getMessage()).getBytes();
                exchange.sendResponseHeaders(500, resp.length);
                exchange.getResponseBody().write(resp);
                exchange.getResponseBody().close();
            }
        });

        server.start();
        System.out.println("Servidor en http://localhost:8080/?n=10");
    }
}

File: golang/version1_soap.go
// http://localhost:9090/?n=10
// go run version1_soap.go

package main

import (
	"fmt"
	"io"
	"net/http"
	"strings"
)

func main() {
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		n := r.URL.Query().Get("n")
		if n == "" {
			fmt.Fprint(w, "Proporciona un número: http://localhost:9090/?n=10")
			return
		}

		soap := `<?xml version="1.0" encoding="utf-8"?>
<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
  <soap:Body>
    <NumberToWords xmlns="http://www.dataaccess.com/webservicesserver/">
      <ubiNum>` + n + `</ubiNum>
    </NumberToWords>
  </soap:Body>
</soap:Envelope>`

		resp, err := http.Post(
			"https://www.dataaccess.com/webservicesserver/NumberConversion.wso",
			"text/xml; charset=utf-8",
			strings.NewReader(soap),
		)
		if err != nil {
			http.Error(w, err.Error(), 500)
			return
		}
		defer resp.Body.Close()

		body, _ := io.ReadAll(resp.Body)
		bodyStr := string(body)
		result := strings.Split(strings.Split(bodyStr, "NumberToWordsResult>")[1], "<")[0]
		fmt.Fprint(w, result)
	})

	fmt.Println("Servidor en http://localhost:9090/?n=10")
	http.ListenAndServe(":9090", nil)
}
